File: scripts/assemble-decode-registry.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Assemble the decode-registry bundle from per-spec JSON sources, byte-exact.
#
# Pretty-printed (2-space indent) and deterministic — specs in sorted-filename
# order, each spec's keys in source order — so a GitHub diff of the bundle reads
# like the diff of the spec sources. There is NO trailing newline: the relay
# serves, signs, and hashes these exact bytes, so a trailing newline (or any
# hand-edit) would break signature verification. Regenerate via this script;
# never hand-edit the generated bundle.

require "json"

KEY_ID = "cosign-registry-2026"

def assemble(specs_dir, out_path)
  specs = Dir.glob(File.join(specs_dir, "*.json")).sort.map { |path| JSON.parse(File.read(path)) }
  bundle = { "schema" => 1, "keyId" => KEY_ID, "specs" => specs }
  File.write(out_path, JSON.pretty_generate(bundle)) # 2-space indent, no trailing newline
end

abort "Usage: assemble-decode-registry.rb <specs-dir> <out.json>" if ARGV.length != 2

assemble(ARGV[0], ARGV[1])

File: scripts/assemble-decode-registry.py
#!/usr/bin/env python3
"""Assemble the decode-registry bundle from per-spec JSON sources, byte-exact.

Emits {"schema":1,"keyId":<KEY_ID>,"specs":[...]} with compact separators and NO
trailing newline, so the served bytes, signed bytes, and hashed bytes are identical.
"""
import json, sys, pathlib

KEY_ID = "cosign-registry-2026"

def main(specs_dir: str, out_path: str) -> None:
    specs = [json.loads(p.read_text()) for p in sorted(pathlib.Path(specs_dir).glob("*.json"))]
    bundle = {"schema": 1, "keyId": KEY_ID, "specs": specs}
    body = json.dumps(bundle, separators=(",", ":"), ensure_ascii=False)
    pathlib.Path(out_path).write_text(body)  # no trailing newline

if __name__ == "__main__":
    if len(sys.argv) != 3:
        sys.exit("Usage: assemble-decode-registry.py <specs-dir> <out.json>")
    main(sys.argv[1], sys.argv[2])
